### ISSUE TITLE
Thoughts on adding direct-links to heading tags within article content?

### ISSUE THREAD

**Issue by chrisdavidmills** [mdn/kuma#7546]
@tannerdolby commented on [Wed Oct 07 2020](https://github.com/mdn/sprints/issues/3757)

## URL(s)
[Content Categories - MDN](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories)

## Request type

- [x] Correction or update

## Details

Hello, MDN team. I'm not so sure this is an issue or problem but something I wanted to ask your team about. 

I originally noticed it on this page [Content Categories - MDN](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories) but I think it applies to all pages within `/docs/Web/`. I wanted to click on the section heading, [Palpable content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Palpable_content) to get a direct link to this section without having to scroll down from [Content Categories](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Main_content_categories). I inspected the page and copied the corresponding id selector `#Palpable_content` and pasted it to the end of the page URL to provide a "direct-link" to that sections content.

I do recognize the "On this page" section containing direct links to main headings of the page in that unordered list which is very helpful. But for sections with more content that require the user to scroll further down the page, this is where I think potentially having direct links on some of the subheadings in long categories could be useful for people sharing links to specific sections within MDN docs.

### Proposed update: 
1. Create direct links for section headings that are further down from the "On this page" direct links. Allowing for a direct link to that section heading to be produced in the browsers address bar upon clicking the `.direct-link` element nested in the heading for fast internal linking. I used the hashtag character as the anchors value (which could also be an SVG if there is a specific icon wanting to be used) within the direct link screen shot I attached below.

The `.direct-link` anchor elements could be fully visible or visually hidden from the screen with good ole `visibility: hidden` or potentially a fancier way of visually hiding using absolute positioning and translate. 

The `.direct-link` element could then become visible upon hovering the heading element and allow for a quick click to update the page URL with a direct link to that specific section heading.

```
<h3 id="Palpable_content">Palpable content <a class="direct-link" href="https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Palpable_content">#</a></h3>
```

## Screenshot with direct link visible for Palpable content heading: (no styling or margins currently applied)
<img width="708" alt="Screen Shot 2020-10-06 at 4 02 18 PM" src="https://user-images.githubusercontent.com/48612525/95269038-6c446f80-07ed-11eb-98e6-11cf0c4c73b1.png">

Let me know your thoughts!

---

@chrisdavidmills commented on [Wed Oct 07 2020](https://github.com/mdn/sprints/issues/3757#issuecomment-704711601)

Hi @tannerdolby !

Yes, I agree this would be nice, and in fact we did have this at some point, but I think it was disabled in the "viewer" front-end because of some kind of performance reason. If you look at the "wiki"/"edit" version of the page (https://wiki.developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories), you'll see the feature active there.

I think it'd be good for us to have this feature sorted out again at some point soon in the final UI, therefore I'm going to move your issue over to the platform (kuma) repo so I can add your voice to the discussion. Thanks for speaking up.

**Comment by peterbe** [comment]
Should we move this to `mdn/yari`?

In Yari, every `<h2>` always gets an ID and every text in the `<h2>` becomes a clickable link that goes to itself.
E.g. https://main.content.dev.mdn.mozit.cloud/en-US/docs/Web/Guide/HTML/Content_categories
We can totally do the same treatment to `<h3>` tags too. It will require a different approach since the `<h2>` is constructed in React whereas the `<h3>` tags are part of the HTML blob that is injected as a HTML string. 

By the way, the inspiration for how our `<h2>` permalinks work came from GitHub Docs ([example](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions)).

**Comment by tannerdolby** [comment]
Hi, Peter. I checked out the [page](https://main.content.dev.mdn.mozit.cloud/en-US/docs/Web/Guide/HTML/Content_categories) in Yari and noticed the `<h2>` headings become a clickable direct link which is very nice. 

If the `<h3>` tags could get the same treatment that would be great. I see, so the `<h2>` tags are JSX and then `<h3>` tags are part of the HTML blob that gets injected. 

Thanks for the link. Yeah, I see how the permalink feature on [Github Docs example](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions ) is similar to those in `mdn/Yari`.

I like the direct permalinks used in Yari inspired by the GitHub Docs example,

<img width="533" alt="direct permalink in Yari" src="https://user-images.githubusercontent.com/48612525/95381135-90ab5500-089c-11eb-9978-71a242187d79.png">

and also the permalink usage Chris mentioned on the ["wiki/edit" version](https://wiki.developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories) in my [previous issue](https://github.com/mdn/sprints/issues/3757) within `mdn/sprints`.

<img width="696" alt="direct permalink within wiki/edit version" src="https://user-images.githubusercontent.com/48612525/95381289-cb14f200-089c-11eb-8f76-3ee2e92d79ba.png">

**Comment by peterbe** [comment]
I certainly like the idea that we rewrite the all `<h3>` tags in the HTML blob to be like the `<h2>` tags in JSX. 
We have the HTML blob parsed with cheerio so it'd be something like this:
```javascript
$('h3').each((i, header) => {
  const $header = $(header)
  if (some reason it's already manually messed with) {
    return
 }
  var id = generateUniqueH3ID($header.text())
  $header.replaceWith($('<a>').attr('href', `#${id}`).text($header.text())
})
```

But let's not jump into this yet. Trying to constrain us to the basic launch of Yari first. Unless you wanna have a try and see what you can prototype up?

**Comment by tannerdolby** [comment]
I'm on board for trying to prototype something up. I had a look at the cheerio documentation and the code you included for the HTML blob example makes sense. 

```
$('h3').each((i, header) => {
  const $header = $(header)
  if (some reason it's already manually messed with) {
    return
 }
  var id = generateUniqueH3ID($header.text())
  $header.replaceWith($('<a>').attr('href', `#${id}`).text($header.text())
})
```

To make sure I'm understanding, we grab the `<h3>` tags with a cheerio selector, `$('h3')`. To iterate over the h3 cheerio objects and create a read only reference $header that is assigned the current header element HTML string, where $(this)=$(header). Next just handle the conditional if for some reason the heading has already been messed with. 

Lastly create the id variable and assign it the generated unique id using `generateUniqueH3ID` with the $header.text() content passed in as a parameter. Then replacing the matched $header content with an `<a>` element and setting its href attribute to `href="#${id}"` using the `id` variable. The link destination or inner content for `<a>` will be the text content from the cheerio object $header since the static method `text` is used, looking something like `<a href="#id">Text from $header</a>` just as a proper direct link should!

I had a look through the `mdn/kuma` repo after forking/cloning and thought maybe using _isHeading(self, token) within the [SectionFilter Class](https://github.com/mdn/kuma/blob/master/kuma/wiki/content.py#L930) in `kuma/wiki/content.py` to check IF the heading has been manually created/messed with, to filter or provide a replacement section content by ID (this might be way unrelated but wanted to see if it had any use).

```
def _isHeading(self, token):
        """Is this token a heading element?"""
        return token["name"] in self.HEADING_TAGS
```
Or maybe something more cheerio like checking if the header had an `<a>` element nested or a certain attr to signal it being manually messed with.
```
if ($header.has(element), $header.has(attr)) {
    var val = $header.text()
    return `Error: Invalid heading ${val}, already messed with`
}
```

I'm happy to try continuing upon what you wrote and or what you suggest if you have the time. I tried finding where the cheerio parsing was going on in `mdn/kuma` but `content.py` was the closest I got to anything HTML related. Maybe if I have a look at the files currently parsing the HTML blobs in the project I will have a better shot at getting something put together.

**Comment by peterbe** [comment]
We're actually moving away from that Python code. The new platform is called [Yari](https://github.com/mdn/yari) and it means it's all `cheerio` and no Python.

**Comment by tannerdolby** [comment]
Ok. Ah yes, I remember you mentioning:

> Should we move this to mdn/yari?

Now that I've had a look at the [Yari](https://github.com/tannerdolby/yari) repo things make more sense. Yari is storing MDN [content](https://github.com/mdn/content) as `index.html` files in `git` that contain the front-matter and metadata to then have those source files built and served as static files. This sort of reminds me of a Static Site generator workflow like 11ty which uses a YAML syntax for the front matter inside opening and closing document seperators `---`. 

The data accessed on [line 11](https://github.com/mdn/content/blob/main/files/en-us/web/html/index.html#L11) of `index.html` uses`{{ variable }}` syntax which is Nunjucks right? (or another templating language?)

Lastly, will the MDN [Content Categories](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories) page  located within [`index.html`](https://github.com/mdn/content/blob/main/files/en-us/web/guide/html/content_categories/index.html) be the same HTML corresponding to the [Yari one](https://main.content.dev.mdn.mozit.cloud/en-US/docs/Web/Guide/HTML/Content_categories)? Now that I know where the Markup is, all I need is to find where the `cheerio` code lives and I can start trying to prototype something up.

**Comment by peterbe** [comment]
You got it! The only thing you didn't get right is that it's not Nunjucks but something else called [`kumascript`](https://github.com/mdn/yari/tree/master/kumascript) which we're not proud of. It was maybe great and necessary many years ago when it was introduced but right now we feel it adds more complexity than it's worth. But here we are. 

Basically, this is what happens:

1. Loading into memory:
```
---
title: 'HTML: HyperText Markup Language'
---
<p>Check this out: {{xref('foo')}} </p>

<h2>Header</h2>
<h3>Sub-header</h3>
<p>Bla bla</p>
```

2. Break it apart into metadata:
```javascript
{title: 'HTML: HyperText Markup Language'}
```
...and the raw HTML:
```
<p>Check this out: {{xref('foo')}} </p>

<h2>Header</h2>
<h3>Sub-header</h3>
<p>Bla bla</p>
```

3. Expand all `kumascript` macros into what we called "rendered html"
```html
<p>Check this out: <a href="/path/to/foo"><code>foo</code></a> </p>

<h2>Header</h2>
<h3>Sub-header</h3>
<p>Bla bla</p>
```

4. Parse the whole thing and break it up into "sections" split by `<h2>` tags:
```javascript
>> console.log(body.sections.length)
2

>> console.log(body.sections[0])
{type: "prose", value: {
  id: null, title: null, content: '<p>Check this out: <a href="/path/to/foo"><code>foo</code></a> </p>'
}}

>> console.log(body.sections[1])
{type: "prose", value: {
  id: "Header", title: "Header", content: '<h3>Sub-header</h3>\n<p>Bla bla</p>'
}}
```

5. Loop over these sections with a React thing that renders them out. This React component treats `'<h3>Sub-header</h3>\n<p>Bla bla</p>'` like any other string and presumes it's safe to inject that HTML string into the DOM without escaping the `<` and `>`. 


What your job would be is to convert the string `<h3>Sub-header</h3>\n<p>Bla bla</p>` into `<h3 id="sub-header-0"><a href="#sub-header-0">Sub-header</a></h3>\n<p>Bla bla</p>`.

Basically, 
1. deserialize from HTML-string to Cheerio object
2. mutate Cheerio object
3. serializer Cheerio object out to HTML-string

Simple!

I think you can look at the function `syntaxHighlight()` or the function `injectNoTranslate()` which is given a Cheerio object (called `$`) which it then mutates in-place. So that might be your ticket to loop over `<h3>` tags like [mentioned above](https://github.com/mdn/kuma/issues/7546#issuecomment-705171042).
